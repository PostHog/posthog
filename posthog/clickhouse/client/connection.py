from contextlib import contextmanager
from enum import Enum
from functools import lru_cache

from clickhouse_driver import Client as SyncClient
from clickhouse_pool import ChPool
from django.conf import settings


class Workload(Enum):
    # Default workload
    DEFAULT = "DEFAULT"
    # Analytics queries, other 'lively' queries
    ONLINE = "ONLINE"
    # Historical exports, other long-running processes where latency is less critical
    OFFLINE = "OFFLINE"


_default_workload = Workload.ONLINE


def get_pool(workload: Workload, team_id=None, readonly=False):
    """
    Returns the right connection pool given a workload.

    Note that the same pool should be returned every call.
    """
    if team_id is not None and str(team_id) in settings.CLICKHOUSE_PER_TEAM_SETTINGS:
        return make_ch_pool(**settings.CLICKHOUSE_PER_TEAM_SETTINGS[str(team_id)])

    # Note that `readonly` does nothing if the relevant vars are not set!
    if readonly and settings.READONLY_CLICKHOUSE_USER is not None and settings.READONLY_CLICKHOUSE_PASSWORD:
        return make_ch_pool(
            user=settings.READONLY_CLICKHOUSE_USER,
            password=settings.READONLY_CLICKHOUSE_PASSWORD,
        )

    if (
        workload == Workload.OFFLINE or workload == Workload.DEFAULT and _default_workload == Workload.OFFLINE
    ) and settings.CLICKHOUSE_OFFLINE_CLUSTER_HOST is not None:
        return make_ch_pool(host=settings.CLICKHOUSE_OFFLINE_CLUSTER_HOST)

    return make_ch_pool()


def default_client():
    """
    Return a bare bones client for use in places where we are only interested in general ClickHouse state
    DO NOT USE THIS FOR QUERYING DATA
    """
    return SyncClient(
        host=settings.CLICKHOUSE_HOST,
        # We set "system" here as we don't necessarily have a "default" database,
        # which is what the clickhouse_driver would use by default. We are
        # assuming that this exists and we have permissions to access it. This
        # feels like a reasonably safe assumption as e.g. we already reference
        # `system.numbers` in multiple places within queries. We also assume
        # access to various other tables e.g. to handle async migrations.
        database="system",
        secure=settings.CLICKHOUSE_SECURE,
        user=settings.CLICKHOUSE_USER,
        password=settings.CLICKHOUSE_PASSWORD,
        ca_certs=settings.CLICKHOUSE_CA,
        verify=settings.CLICKHOUSE_VERIFY,
    )


@lru_cache(maxsize=None)
def make_ch_pool(**overrides) -> ChPool:
    kwargs = {
        "host": settings.CLICKHOUSE_HOST,
        "database": settings.CLICKHOUSE_DATABASE,
        "secure": settings.CLICKHOUSE_SECURE,
        "user": settings.CLICKHOUSE_USER,
        "password": settings.CLICKHOUSE_PASSWORD,
        "ca_certs": settings.CLICKHOUSE_CA,
        "verify": settings.CLICKHOUSE_VERIFY,
        "connections_min": settings.CLICKHOUSE_CONN_POOL_MIN,
        "connections_max": settings.CLICKHOUSE_CONN_POOL_MAX,
        "settings": {"mutations_sync": "1"} if settings.TEST else {},
        # Without this, OPTIMIZE table and other queries will regularly run into timeouts
        "send_receive_timeout": 30 if settings.TEST else 999_999_999,
        **overrides,
    }

    return ChPool(**kwargs)


@contextmanager
def set_default_clickhouse_workload_type(workload: Workload):
    global _default_workload

    _default_workload = workload


ch_pool = get_pool(workload=Workload.ONLINE)
