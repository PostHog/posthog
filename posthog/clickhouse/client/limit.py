import time
from functools import wraps
from typing import Optional
from collections.abc import Callable

from celery import current_task
from prometheus_client import Counter

from posthog import redis

CONCURRENT_TASKS_LIMIT_EXCEEDED_COUNTER = Counter(
    "posthog_celery_task_concurrency_limit_exceeded",
    "Number of times a Celery task exceeded the concurrency limit",
    ["task_name", "limit"],
)

# Lua script for atomic check, remove expired if limit hit, and increment with TTL
lua_script = """
local key = KEYS[1]
local current_time = tonumber(ARGV[1])
local task_id = ARGV[2]
local max_concurrent_tasks = tonumber(ARGV[3])
local ttl = tonumber(ARGV[4])
local expiration_time = current_time + ttl

-- Check the number of current running tasks
local running_tasks_count = redis.call('ZCARD', key)
if running_tasks_count >= max_concurrent_tasks then
    -- Remove expired tasks if limit is hit
    redis.call('ZREMRANGEBYSCORE', key, '-inf', current_time)
    running_tasks_count = redis.call('ZCARD', key)
    if running_tasks_count >= max_concurrent_tasks then
        return 0
    end
end

-- Add the new task with its expiration time
redis.call('ZADD', key, expiration_time, task_id)
return 1
"""


class CeleryConcurrencyLimitExceeded(Exception):
    pass


def limit_concurrency(max_concurrent_tasks: int, key: Optional[Callable] = None, ttl: int = 60 * 15) -> Callable:
    def decorator(task_func):
        @wraps(task_func)
        def wrapper(*args, **kwargs):
            task_name = current_task.name
            redis_client = redis.get_client()
            running_tasks_key = f"celery_running_tasks:{task_name}"
            if key:
                dynamic_key = key(*args, **kwargs)
                running_tasks_key = f"{running_tasks_key}:{dynamic_key}"
            else:
                dynamic_key = None
            task_id = f"{task_name}:{current_task.request.id}"
            current_time = int(time.time())

            # Atomically check, remove expired if limit hit, and add the new task
            if (
                redis_client.eval(lua_script, 1, running_tasks_key, current_time, task_id, max_concurrent_tasks, ttl)
                == 0
            ):
                CONCURRENT_TASKS_LIMIT_EXCEEDED_COUNTER.labels(task_name=task_name, limit=max_concurrent_tasks).inc()

                raise CeleryConcurrencyLimitExceeded(
                    f"Exceeded maximum concurrent tasks limit: {max_concurrent_tasks} for key: {dynamic_key}"
                )

            try:
                # Execute the task
                return task_func(*args, **kwargs)
            finally:
                # Remove the task ID from the sorted set when the task finishes
                redis_client.zrem(running_tasks_key, task_id)

        return wrapper

    return decorator
