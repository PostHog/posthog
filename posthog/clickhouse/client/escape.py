# Methods used for rendering a ClickHouse query, given a template string and a
# set of parameters.
#
# This uses the `escape_param` function from the `clickhouse-driver` package,
# but passes an empty `Context` object to it. Prior to
# https://github.com/mymarilyn/clickhouse-driver/commit/87090902f0270ed51a0b6754d5cbf0dc8544ec4b
# the `escape_param` function didn't take a `Context` object. As of
# `clickhouse-driver` 0.2.4 all it uses the context for is to determine the
# "server" timezone, so passing an empty context maintains the existing
# behaviour of `clickhouse-driver` 0.2.1, the version we were previously using.
#
# This is of course a bit of a hack but we want to be able to render queries
# without the need of having a connection, which seems like a reasonable thing
# to be able to do. Having a dependency on a connection to render a query is a
# little over the top.
#
# NOTE: this change is necessary because the `clickhouse-driver` package up to
# 0.2.3 uses an invalid `python_requires` in it's `setup.py` at least for
# recent versions of setuptools. This was highlighted as a consequence of
# upgrading to Python 3.10. See
# https://github.com/mymarilyn/clickhouse-driver/pull/291 for further context.


from typing import Any

from clickhouse_driver.connection import ServerInfo
from clickhouse_driver.context import Context
from clickhouse_driver.util.escape import escape_param


def substitute_params(query, params):
    """
    This is a copy of clickhouse-driver's `substitute_params` function without
    the dependency that you need to connect to the server before you can escape
    params. There was a bug in which we were trying to substitute params before
    the connection was established, which caused the query to fail. Presumably
    this was on initial worker startup only.

    It seems somewhat unusual that you need to connect to the server before
    you can escape params, so we're just going to copy the function here
    and remove that dependency.

    See
    https://github.com/mymarilyn/clickhouse-driver/blob/87090902f0270ed51a0b6754d5cbf0dc8544ec4b/clickhouse_driver/client.py#L593
    for the original function.
    """
    if not isinstance(params, dict):
        raise ValueError("Parameters are expected in dict form")

    escaped = escape_params(params)
    return query % escaped


def escape_params(params):
    """
    This is a copy of clickhouse-driver's `escape_params` function without the
    dependency that you need to connect to the server before you can escape
    params.

    See
    https://github.com/mymarilyn/clickhouse-driver/blob/87090902f0270ed51a0b6754d5cbf0dc8544ec4b/clickhouse_driver/util/escape.py#L60
    for the original function.
    """
    escaped = {}

    for key, value in params.items():
        escaped[key] = escape_param_for_clickhouse(value)

    return escaped


def escape_param_for_clickhouse(param: Any) -> str:
    """
    This is a wrapper around the `escape_param` function from the
    `clickhouse-driver` package, but passes a placeholder `Context` object to it
    just such that it can run. The only value that the real `escape_param` uses
    from the context is the server timezone. We assume that the server timezone
    is UTC.

    See
    https://github.com/mymarilyn/clickhouse-driver/blob/87090902f0270ed51a0b6754d5cbf0dc8544ec4b/clickhouse_driver/util/escape.py#L31
    for the wrapped function.
    """
    context = Context()
    context.server_info = ServerInfo(
        name="placeholder server_info value",
        version_major="placeholder server_info value",
        version_minor="placeholder server_info value",
        version_patch="placeholder server_info value",
        revision="placeholder server_info value",
        display_name="placeholder server_info value",
        timezone="UTC",
    )
    return escape_param(param, context=context)
