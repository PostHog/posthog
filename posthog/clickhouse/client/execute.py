import json
import threading
import types
from contextlib import contextmanager
from functools import lru_cache
from time import perf_counter
from typing import Any, Dict, List, Optional, Sequence, Union

import sqlparse
from clickhouse_driver import Client as SyncClient
from django.conf import settings as app_settings
from statshog.defaults.django import statsd

from posthog.clickhouse.client.connection import Workload, get_pool
from posthog.clickhouse.client.escape import substitute_params
from posthog.clickhouse.query_tagging import get_query_tag_value, get_query_tags
from posthog.errors import wrap_query_error
from posthog.settings import TEST
from posthog.utils import generate_short_id, patchable

InsertParams = Union[list, tuple, types.GeneratorType]
NonInsertParams = Dict[str, Any]
QueryArgs = Optional[Union[InsertParams, NonInsertParams]]

thread_local_storage = threading.local()

# As of CH 22.8 - more algorithms have been added on newer versions
CLICKHOUSE_SUPPORTED_JOIN_ALGORITHMS = [
    "default",
    "hash",
    "parallel_hash",
    "direct",
    "full_sorting_merge",
    "partial_merge",
    "auto",
]

is_invalid_algorithm = lambda algo: algo not in CLICKHOUSE_SUPPORTED_JOIN_ALGORITHMS


@lru_cache(maxsize=1)
def default_settings() -> Dict:
    return {
        "join_algorithm": "direct,parallel_hash",
        "distributed_replica_max_ignored_errors": 1000,
    }


@lru_cache(maxsize=1)
def clickhouse_at_least_228() -> bool:
    from posthog.version_requirement import ServiceVersionRequirement

    is_ch_version_228_or_above, _ = ServiceVersionRequirement(
        service="clickhouse", supported_version=">=22.8.0"
    ).is_service_in_accepted_version()

    return is_ch_version_228_or_above


def validated_client_query_id() -> Optional[str]:
    client_query_id = get_query_tag_value("client_query_id")
    client_query_team_id = get_query_tag_value("team_id")

    if client_query_id and not client_query_team_id:
        raise Exception("Query needs to have a team_id arg if you've passed client_query_id")
    random_id = generate_short_id()
    return f"{client_query_team_id}_{client_query_id}_{random_id}"


@patchable
def sync_execute(
    query,
    args=None,
    settings=None,
    with_column_types=False,
    flush=True,
    *,
    workload: Workload = Workload.DEFAULT,
    team_id: Optional[int] = None,
    readonly=False,
):
    if TEST and flush:
        try:
            from posthog.test.base import flush_persons_and_events

            flush_persons_and_events()
        except ModuleNotFoundError:  # when we run plugin server tests it tries to run above, ignore
            pass

    with get_pool(workload, team_id, readonly).get_client() as client:
        start_time = perf_counter()

        prepared_sql, prepared_args, tags = _prepare_query(client=client, query=query, args=args, workload=workload)
        query_id = validated_client_query_id()
        core_settings = {**default_settings(), **(settings or {})}
        tags["query_settings"] = core_settings
        settings = {
            **core_settings,
            "log_comment": json.dumps(tags, separators=(",", ":")),
        }
        try:
            result = client.execute(
                prepared_sql,
                params=prepared_args,
                settings=settings,
                with_column_types=with_column_types,
                query_id=query_id,
            )
        except Exception as err:
            err = wrap_query_error(err)
            statsd.incr(
                "clickhouse_sync_execution_failure",
                tags={"failed": True, "reason": type(err).__name__},
            )

            raise err
        finally:
            execution_time = perf_counter() - start_time

            statsd.timing("clickhouse_sync_execution_time", execution_time * 1000.0)

            if query_counter := getattr(thread_local_storage, "query_counter", None):
                query_counter.total_query_time += execution_time

            if app_settings.SHELL_PLUS_PRINT_SQL:
                print("Execution time: %.6fs" % (execution_time,))  # noqa T201
    return result


def query_with_columns(
    query: str,
    args: Optional[QueryArgs] = None,
    columns_to_remove: Optional[Sequence[str]] = None,
    columns_to_rename: Optional[Dict[str, str]] = None,
    *,
    workload: Workload = Workload.DEFAULT,
    team_id: Optional[int] = None,
) -> List[Dict]:
    if columns_to_remove is None:
        columns_to_remove = []
    if columns_to_rename is None:
        columns_to_rename = {}
    metrics, types = sync_execute(query, args, with_column_types=True, workload=workload, team_id=team_id)
    type_names = [key for key, _type in types]

    rows = []
    for row in metrics:
        result = {}
        for type_name, value in zip(type_names, row):
            if type_name not in columns_to_remove:
                result[columns_to_rename.get(type_name, type_name)] = value

        rows.append(result)

    return rows


@patchable
def _prepare_query(
    client: SyncClient,
    query: str,
    args: QueryArgs,
    workload: Workload = Workload.DEFAULT,
):
    """
    Given a string query with placeholders we do one of two things:

        1. for a insert query we just format, and remove comments
        2. for non-insert queries, we return the sql with placeholders
        evaluated with the contents of `args`

    We also return `tags` which contains some detail around the context
    within which the query was executed e.g. the django view name

    NOTE: `client.execute` would normally handle substitution, but
    because we want to strip the comments to make it easier to copy
    and past queries from the `system.query_log` easily with metabase
    (metabase doesn't show new lines, so with comments, you can't get
    a working query without exporting to csv or similar), we need to
    do it manually.

    We only want to try to substitue for SELECT queries, which
    clickhouse_driver at this moment in time decides based on the
    below predicate.
    """
    prepared_args: Any = QueryArgs
    if isinstance(args, (list, tuple, types.GeneratorType)):
        # If we get one of these it means we have an insert, let the clickhouse
        # client handle substitution here.
        rendered_sql = query
        prepared_args = args
    elif not args:
        # If `args` is not truthy then make prepared_args `None`, which the
        # clickhouse client uses to signal no substitution is desired. Expected
        # args balue are `None` or `{}` for instance
        rendered_sql = query
        prepared_args = None
    else:
        # Else perform the substitution so we can perform operations on the raw
        # non-templated SQL
        rendered_sql = substitute_params(query, args)
        prepared_args = None

    formatted_sql = sqlparse.format(rendered_sql, strip_comments=True)
    annotated_sql, tags = _annotate_tagged_query(formatted_sql, workload)

    if app_settings.SHELL_PLUS_PRINT_SQL:
        print()  # noqa T201
        print(format_sql(formatted_sql))  # noqa T201

    return annotated_sql, prepared_args, tags


def _annotate_tagged_query(query, workload):
    """
    Adds in a /* */ so we can look in clickhouses `system.query_log`
    to easily marry up to the generating code.
    """
    tags = {**get_query_tags(), "workload": str(workload)}
    # Annotate the query with information on the request/task
    if "kind" in tags:
        user_id = f" user_id:{tags['user_id']}" if "user_id" in tags else ""
        query = f"/*{user_id} {tags.get('kind')}:{tags.get('id', '').replace('/', '_')} */ {query}"

    return query, tags


def format_sql(rendered_sql, colorize=True):
    formatted_sql = sqlparse.format(rendered_sql, reindent_aligned=True)
    if colorize:
        try:
            import pygments.formatters
            import pygments.lexers

            return pygments.highlight(
                formatted_sql,
                pygments.lexers.get_lexer_by_name("sql"),
                pygments.formatters.TerminalFormatter(),
            )
        except:
            pass

    return formatted_sql


@contextmanager
def clickhouse_query_counter(query_counter):
    thread_local_storage.query_counter = query_counter
    yield
    thread_local_storage.query_counter = None
