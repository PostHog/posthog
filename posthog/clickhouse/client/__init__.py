from posthog.clickhouse.client.execute import query_with_columns, sync_execute
from posthog.clickhouse.client.execute_async import execute_with_progress

__all__ = [
    "sync_execute",
    "query_with_columns",
    "execute_with_progress",
]
