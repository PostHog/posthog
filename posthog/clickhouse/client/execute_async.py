import hashlib
import json
import time
from dataclasses import asdict as dataclass_asdict
from dataclasses import dataclass
from time import perf_counter
from typing import Any, Optional

from posthog import celery
from clickhouse_driver import Client as SyncClient
from django.conf import settings as app_settings
from statshog.defaults.django import statsd

from posthog import redis
from posthog.celery import enqueue_clickhouse_execute_with_progress
from posthog.clickhouse.client.execute import _prepare_query
from posthog.errors import wrap_query_error
from posthog.settings import (
    CLICKHOUSE_CA,
    CLICKHOUSE_DATABASE,
    CLICKHOUSE_HOST,
    CLICKHOUSE_PASSWORD,
    CLICKHOUSE_SECURE,
    CLICKHOUSE_USER,
    CLICKHOUSE_VERIFY,
)

REDIS_STATUS_TTL = 600  # 10 minutes


@dataclass
class QueryStatus:
    team_id: int
    num_rows: float = 0
    total_rows: float = 0
    error: bool = False
    complete: bool = False
    error_message: str = ""
    results: Any = None
    start_time: Optional[float] = None
    end_time: Optional[float] = None
    task_id: Optional[str] = None


def generate_redis_results_key(query_id):
    REDIS_KEY_PREFIX_ASYNC_RESULTS = "query_with_progress"
    key = f"{REDIS_KEY_PREFIX_ASYNC_RESULTS}:{query_id}"
    return key


def execute_with_progress(
    team_id,
    query_id,
    query,
    args=None,
    settings=None,
    with_column_types=False,
    update_freq=0.2,
    task_id=None,
):
    """
    Kick off query with progress reporting
    Iterate over the progress status
    Save status to redis
    Once complete save results to redis
    """

    key = generate_redis_results_key(query_id)
    ch_client = SyncClient(
        host=CLICKHOUSE_HOST,
        database=CLICKHOUSE_DATABASE,
        secure=CLICKHOUSE_SECURE,
        user=CLICKHOUSE_USER,
        password=CLICKHOUSE_PASSWORD,
        ca_certs=CLICKHOUSE_CA,
        verify=CLICKHOUSE_VERIFY,
        settings={"max_result_rows": "10000"},
    )
    redis_client = redis.get_client()

    start_time = perf_counter()

    prepared_sql, prepared_args, tags = _prepare_query(client=ch_client, query=query, args=args)

    query_status = QueryStatus(team_id, task_id=task_id)

    start_time = time.time()

    try:
        progress = ch_client.execute_with_progress(
            prepared_sql,
            params=prepared_args,
            settings=settings,
            with_column_types=with_column_types,
        )
        for num_rows, total_rows in progress:
            query_status = QueryStatus(
                team_id=team_id,
                num_rows=num_rows,
                total_rows=total_rows,
                complete=False,
                error=False,
                error_message="",
                results=None,
                start_time=start_time,
                task_id=task_id,
            )
            redis_client.set(key, json.dumps(dataclass_asdict(query_status)), ex=REDIS_STATUS_TTL)
            time.sleep(update_freq)
        else:
            rv = progress.get_result()
            query_status = QueryStatus(
                team_id=team_id,
                num_rows=query_status.num_rows,
                total_rows=query_status.total_rows,
                complete=True,
                error=False,
                start_time=query_status.start_time,
                end_time=time.time(),
                error_message="",
                results=rv,
                task_id=task_id,
            )
            redis_client.set(key, json.dumps(dataclass_asdict(query_status)), ex=REDIS_STATUS_TTL)

    except Exception as err:
        err = wrap_query_error(err)
        tags["failed"] = True
        tags["reason"] = type(err).__name__
        statsd.incr("clickhouse_sync_execution_failure")
        query_status = QueryStatus(
            team_id=team_id,
            num_rows=query_status.num_rows,
            total_rows=query_status.total_rows,
            complete=False,
            error=True,
            start_time=query_status.start_time,
            end_time=time.time(),
            error_message=str(err),
            results=None,
            task_id=task_id,
        )
        redis_client.set(key, json.dumps(dataclass_asdict(query_status)), ex=REDIS_STATUS_TTL)

        raise err
    finally:
        ch_client.disconnect()

        execution_time = perf_counter() - start_time

        statsd.timing("clickhouse_sync_execution_time", execution_time * 1000.0)

        if app_settings.SHELL_PLUS_PRINT_SQL:
            print("Execution time: %.6fs" % (execution_time,))  # noqa T201


def enqueue_execute_with_progress(
    team_id,
    query,
    args=None,
    settings=None,
    with_column_types=False,
    bypass_celery=False,
    query_id=None,
    force=False,
):
    if not query_id:
        query_id = _query_hash(query, team_id, args)
    key = generate_redis_results_key(query_id)
    redis_client = redis.get_client()

    if force:
        # If we want to force rerun of this query we need to
        # 1) Get the current status from redis
        task_str = redis_client.get(key)
        if task_str:
            # if the status exists in redis we need to tell celery to kill the job
            task_str = task_str.decode("utf-8")
            query_task = QueryStatus(**json.loads(task_str))
            # Instruct celery to revoke task and terminate if running
            celery.app.control.revoke(query_task.task_id, terminate=True)
            # Then we need to make redis forget about this job entirely
            # and continue as normal. As if we never saw this query before
            redis_client.delete(key)

    if redis_client.get(key):
        # If we've seen this query before return the query_id and don't resubmit it.
        return query_id

    # Immediately set status so we don't have race with celery
    query_status = QueryStatus(team_id=team_id, start_time=time.time())
    redis_client.set(key, json.dumps(dataclass_asdict(query_status)), ex=REDIS_STATUS_TTL)

    if bypass_celery:
        # Call directly ( for testing )
        enqueue_clickhouse_execute_with_progress(team_id, query_id, query, args, settings, with_column_types)
    else:
        enqueue_clickhouse_execute_with_progress.delay(team_id, query_id, query, args, settings, with_column_types)

    return query_id


def get_status_or_results(team_id, query_id):
    """
    Returns QueryStatus data class
    QueryStatus data class contains either:
    Current status of running query
    Results of completed query
    Error payload of failed query
    """
    redis_client = redis.get_client()
    key = generate_redis_results_key(query_id)
    try:
        byte_results = redis_client.get(key)
        if byte_results:
            str_results = byte_results.decode("utf-8")
        else:
            return QueryStatus(team_id, error=True, error_message="Query is unknown to backend")
        query_status = QueryStatus(**json.loads(str_results))
        if query_status.team_id != team_id:
            raise Exception("Requesting team is not executing team")
    except Exception as e:
        query_status = QueryStatus(team_id, error=True, error_message=str(e))
    return query_status


def _query_hash(query: str, team_id: int, args: Any) -> str:
    """
    Takes a query and returns a hex encoded hash of the query and args
    """
    if args:
        key = hashlib.md5((str(team_id) + query + json.dumps(args)).encode("utf-8")).hexdigest()
    else:
        key = hashlib.md5((str(team_id) + query).encode("utf-8")).hexdigest()
    return key
