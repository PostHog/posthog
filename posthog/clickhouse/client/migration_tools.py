from infi.clickhouse_orm import migrations

from posthog.clickhouse.client.execute import sync_execute


def run_sql_with_exceptions(sql: str, settings={}):
    """
    migrations.RunSQL does not raise exceptions, so we need to wrap it in a function that does.
    """

    def run_sql(database):
        sync_execute(sql, settings=settings)

    return migrations.RunPython(run_sql)
