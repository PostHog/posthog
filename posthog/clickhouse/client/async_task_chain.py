import threading
import typing
from contextlib import contextmanager
from functools import partial

from celery import chain
from celery.canvas import Signature
from django.db import transaction

from posthog.schema import QueryStatus

if typing.TYPE_CHECKING:
    from posthog.clickhouse.client.execute_async import QueryStatusManager

# Create a thread-local storage for task chains
_thread_locals = threading.local()


def kick_off_task(
    manager: "QueryStatusManager",
    query_status: QueryStatus,
    task_signature: Signature,
) -> None:
    task = task_signature.apply_async()
    query_status.task_id = task.id
    manager.store_query_status(query_status)


def get_task_chain() -> list[tuple[Signature, "QueryStatusManager", QueryStatus]]:
    """
    Retrieves the task chain from thread-local storage.
    """
    if not hasattr(_thread_locals, "task_chain"):
        _thread_locals.task_chain = []
    return _thread_locals.task_chain


def set_in_context(value) -> None:
    """
    Sets the in_context flag in thread-local storage.
    """
    _thread_locals.in_context = value


def is_in_context() -> bool:
    """
    Checks if the in_context flag is set in thread-local storage.
    """
    return getattr(_thread_locals, "in_context", False)


def add_task_to_chain(task_signature: Signature, manager: "QueryStatusManager", query_status: QueryStatus) -> None:
    """
    Adds a task to the chain. If not in context, registers the task with on_commit directly.
    """

    if is_in_context():
        task_chain = get_task_chain()
        task_chain.append((task_signature, manager, query_status))
    else:
        transaction.on_commit(partial(kick_off_task, manager, query_status, task_signature))


def execute_task_chain() -> None:
    """
    Executes the task chain after the transaction is committed.
    """
    task_chain = get_task_chain()
    if task_chain:
        chained_tasks = chain(*[args[0] for args in task_chain])
        result = chained_tasks.apply_async()

        for args in task_chain:
            args[2].task_id = result.id
            args[1].store_query_status(args[2])

        _thread_locals.task_chain = []


@contextmanager
def task_chain_context() -> typing.Iterator[None]:
    set_in_context(True)
    try:
        yield
    finally:
        set_in_context(False)
        transaction.on_commit(execute_task_chain)
