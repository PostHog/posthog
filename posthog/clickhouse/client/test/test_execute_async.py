import uuid
from unittest.mock import patch

from django.test import TestCase

from posthog.clickhouse.client import execute_async as client
from posthog.client import sync_execute
from posthog.hogql.errors import HogQLException
from posthog.models import Organization, Team
from posthog.test.base import ClickhouseTestMixin, snapshot_clickhouse_queries


def build_query(sql):
    return {
        "kind": "HogQLQuery",
        "query": sql,
    }


class ClickhouseClientTestCase(TestCase, ClickhouseTestMixin):
    def setUp(self):
        self.organization = Organization.objects.create(name="test")
        self.team = Team.objects.create(organization=self.organization)
        self.team_id = self.team.pk

    @snapshot_clickhouse_queries
    def test_async_query_client(self):
        query = build_query("SELECT 1+1")
        team_id = self.team_id
        query_id = client.enqueue_process_query_task(team_id, query, bypass_celery=True).id
        result = client.get_query_status(team_id, query_id)
        self.assertFalse(result.error, result.error_message)
        self.assertTrue(result.complete)
        self.assertEqual(result.results["results"], [[2]])

    def test_async_query_client_errors(self):
        query = build_query("SELECT WOW SUCH DATA FROM NOWHERE THIS WILL CERTAINLY WORK")
        self.assertRaises(
            HogQLException,
            client.enqueue_process_query_task,
            **{"team_id": (self.team_id), "query_json": query, "bypass_celery": True},
        )
        query_id = uuid.uuid4().hex
        try:
            client.enqueue_process_query_task(self.team_id, query, query_id=query_id, bypass_celery=True)
        except Exception:
            pass

        result = client.get_query_status(self.team_id, query_id)
        self.assertTrue(result.error)
        self.assertRegex(result.error_message, "Unknown table")

    def test_async_query_client_uuid(self):
        query = build_query("SELECT toUUID('00000000-0000-0000-0000-000000000000')")
        team_id = self.team_id
        query_id = client.enqueue_process_query_task(team_id, query, bypass_celery=True).id
        result = client.get_query_status(team_id, query_id)
        self.assertFalse(result.error, result.error_message)
        self.assertTrue(result.complete)
        self.assertEqual(result.results["results"], [["00000000-0000-0000-0000-000000000000"]])

    def test_async_query_client_does_not_leak(self):
        query = build_query("SELECT 1+1")
        team_id = self.team_id
        wrong_team = 5
        query_id = client.enqueue_process_query_task(team_id, query, bypass_celery=True).id

        try:
            client.get_query_status(wrong_team, query_id)
        except Exception as e:
            self.assertEqual(str(e), f"Query {query_id} not found for team {wrong_team}")

    @patch("posthog.clickhouse.client.execute_async.process_query_task")
    def test_async_query_client_is_lazy(self, execute_sync_mock):
        query = build_query("SELECT 4 + 4")
        query_id = uuid.uuid4().hex
        team_id = self.team_id
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True)

        # Try the same query again
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True)

        # Try the same query again (for good measure!)
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True)

        # Assert that we only called clickhouse once
        execute_sync_mock.assert_called_once()

    @patch("posthog.clickhouse.client.execute_async.process_query_task")
    def test_async_query_client_is_lazy_but_not_too_lazy(self, execute_sync_mock):
        query = build_query("SELECT 8 + 8")
        query_id = uuid.uuid4().hex
        team_id = self.team_id
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True)

        # Try the same query again, but with force
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True, force=True)

        # Try the same query again (for good measure!)
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True)

        # Assert that we called clickhouse twice
        self.assertEqual(execute_sync_mock.call_count, 2)

    @patch("posthog.clickhouse.client.execute_async.process_query_task")
    def test_async_query_client_manual_query_uuid(self, execute_sync_mock):
        # This is a unique test because technically in the test pattern `SELECT 8 + 8` is already
        # in redis. This tests to make sure it is treated as a unique run of that query
        query = build_query("SELECT 8 + 8")
        team_id = self.team_id
        query_id = "I'm so unique"
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True)

        # Try the same query again, but with force
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True, force=True)

        # Try the same query again (for good measure!)
        client.enqueue_process_query_task(team_id, query, query_id=query_id, bypass_celery=True)

        # Assert that we called clickhouse twice
        self.assertEqual(execute_sync_mock.call_count, 2)

    def test_client_strips_comments_from_request(self):
        """
        To ensure we can easily copy queries from `system.query_log` in e.g.
        Metabase, we strip comments from the query we send. Metabase doesn't
        display multilined output.

        See https://github.com/metabase/metabase/issues/14253

        Note I'm not really testing much complexity, I trust that those will
        come out as failures in other tests.
        """
        from posthog.clickhouse.query_tagging import tag_queries

        # First add in the request information that should be added to the sql.
        # We check this to make sure it is not removed by the comment stripping
        with self.capture_select_queries() as sqls:
            tag_queries(kind="request", id="1")
            sync_execute(
                query="""
                    -- this request returns 1
                    SELECT 1
                """
            )
            self.assertEqual(len(sqls), 1)
            first_query = sqls[0]
            self.assertIn(f"SELECT 1", first_query)
            self.assertNotIn("this request returns", first_query)

            # Make sure it still includes the "annotation" comment that includes
            # request routing information for debugging purposes
            self.assertIn("/* request:1 */", first_query)
