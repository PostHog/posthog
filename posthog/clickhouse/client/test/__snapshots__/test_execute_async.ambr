# name: ClickhouseClientTestCase.test_async_query_client
  '
  SELECT plus(1, 1)
  LIMIT 10000 SETTINGS readonly=2,
                       max_execution_time=600,
                       allow_experimental_object_type=1
  '
---
